### PR TITLE
Adds underlines to differences in failing test strings

### PR DESCRIPTION
**Pull request by orta**
Adds an underline to any parts of the string which are different between the `actual` and `expected` when a text diff is raised in tests:

Screen Shot 2020-01-24 at 12 51 38 PM

### CODE REVIEW

File: src/harness/fourslashImpl.ts [hunk 1]
```diff
@@ -4077,10 +4077,57 @@ ${code}
...
-        return `\n${expectMsg}:\n${expectedString}\n\n${actualMsg}:\n${actualString}`;
+        return `\n${expectMsg}:\n${expectedString}\n\n${actualMsg}:\n${highlightDifferenceBetweenStrings(expected, actualString)}`;
     }
 
     function templateToRegExp(template: string) {
         return new RegExp(`^${ts.regExpEscape(template).replace(/\\\{\d+\\\}/g, ".*?")}$`);
     }
+
+    function rangesOfDiffBetweenTwoStrings(source: string, target: string) {
+        const ranges = [] as { start: number; length: number }[];
```
Comment (sandersn) [inline src/harness/fourslashImpl.ts]: you wouldn't need this type annotation if you would inline `addToIndex` to the body of the for loop...

probably not worth the stylistic change though.

File: src/harness/fourslashImpl.ts [hunk 2]
```diff
@@ -4077,10 +4077,57 @@ ${code}
...
+            else {
+                ranges.push({ start: index - 1, length: 1 });
+            }
+          }
+          else {
+                ranges.push({ start: index - 1, length: 1 });
+          }
+        };
+
+        for (let index = 0; index < Math.max(source.length, target.length); index++) {
```
Comment (sandersn) [inline src/harness/fourslashImpl.ts]: I'm not convinced this algorithm is always going to give correct and pretty ranges, but it probably doesn't matter too much.

File: src/harness/fourslashImpl.ts [hunk 3]
```diff
@@ -4077,10 +4077,57 @@ ${code}
...
+        return ranges;
+      }
+
+      // Adds an _ when the source string and the target string have a whitespace difference
+      function highlightDifferenceBetweenStrings(source: string, target: string) {
+        const ranges = rangesOfDiffBetweenTwoStrings(source, target);
+        let emTarget = target;
+        ranges.forEach((range, index) => {
+            const lhs = `\x1b[4m`;
+            const rhs = `\x1b[0m\x1b[31m`;
```
Comment (sandersn) [inline src/harness/fourslashImpl.ts]: just want to double-check, this code is only called from fourslash, so is very very likely to be going to a terminal, right?

**Comment by sandersn** [comment]
@orta this is ready to merge, right?

**Comment by orta** [comment]
Let's give it a shot - if it's annoying for folks it can be easily reverted 👍

**Pull request merged**